### PR TITLE
Adds telemetry for browsing Deprecated and Vulnerable Packages in PM UI

### PR DESCRIPTION
**Pull request by dominoFire**
## Bug

Fixes: https://github.com/NuGet/Client.Engineering/issues/1123

Regression? No Last working version:

## Description

- Adds `vs/nuget/hyperlinkclicked` event, which is triggered when certain PM UI Hyperlink controls are clicked. It also includes info for deprecation message type (with or without alternative package ID).
- Modifies `vs/nuget/searchselection` event to distinguish whether the selected package is vulnerable/deprecated or not, and, in case of a deprecated package, detemine if an alternative package is provided.
- Modifies parameter binding in related deprecation Hyperlink controls using a `IMultiValueConverter`.
- Updates xmldocs
- Unifies package Id normalization to guarantee event corretation between `vs/nuget/searchselection`, `vs/nuget/hyperlinkclicked` and `vs/nuget/nugetaction`

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added: 2 new tests added
  - ~[ ] Test exception~
  - ~[ ] N/A~

- **Documentation**
  - ~[ ] Documentation PR or issue filled~
  - [x] N/A: Engineering changes. No feature is modified

## Validation
 
Telemetry Inspection `vs/nuget/searchselection`

image

Telemetry Inspection `vs/nuget/hyperlinkclicked`

image

### CODE REVIEW

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/HyperlinkType.cs
Comment (dominoFire) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/HyperlinkType.cs:9]: Not all hyperlinks are tracked. A follow up issue is filed to complete tracking of all URIs. See https://github.com/NuGet/Client.Engineering/issues/1124

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/HyperlinkClickedTelemetryEvent.cs
Comment (loic-sharma) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/HyperlinkClickedTelemetryEvent.cs:28]: Should we add information about which source the package is from too? I realize this information is captured in previously emitted events, but, joins in Visual Studio telemetry are expensive on large data sets.

Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/HyperlinkClickedTelemetryEvent.cs:28]: Why it would be useful ? Which questions would you like to answer with feed information ?

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs
Comment (heng-liu) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs:77]: I'm new to the combination of [Constant pattern](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#constant-pattern) and [Declaration and type patterns](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/patterns#declaration-and-type-patterns). Is there any language version requirement to use this combination? Thanks!

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/HyperlinkClickedTelemetryEvent.cs
Comment (loic-sharma) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/HyperlinkClickedTelemetryEvent.cs:28]: From my understanding the deprecation link will search for `packageid:X`, which may not work on all package sources. This telemetry would let us see if this problem is worth solving.

Is there any reasons to be conservative with what data we add to VS telemetry events? Given the slow release cadence, it seems better to add what may be useful instead of only what we know we need... but I don't know VS's best practices guideline in this area :)

**Review by heng-liu — APPROVED** [review]
LGTM from GDPR perspective. 
Pls try to get another approval before merging. Pls raise follow-up issues if needed since it's kind of urgent to get it in.

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/HyperlinkClickedTelemetryEvent.cs
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Telemetry/HyperlinkClickedTelemetryEvent.cs:28]: Sounds good.

I just want to make sure I'm using the right amount of telemetry, and, I have planned a second iteration of this telemetry for next VS Preview release.

Also, deprecation metadata is gathered from the active feed the user has selected in PM UI. In the case of 'All' sources, the first NuGet feed/source that replies with deprecation metadata is shown to the user. We currently don't track deprecation metadata source feed origin. So, there are two options:

- Either we just track current selected feed + whether all sources contains a non-nuget.org remote feed (most feasible, from my point of view)
- Make the whole metadata tracking from feed to hyperlink telemetry

I will add this telemetry in https://github.com/NuGet/Client.Engineering/issues/1124

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs:77]: It's a new feature in C# 9 for conjuntive type match patterns. See examples of https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-9#pattern-matching-enhancements

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Utility/UIUtility.cs]: Add `using ContractItemFilter = NuGet.VisualStudio.Internal.Contracts.ItemFilter;` and reference here.

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs:64]:
```suggestion
        /// Handles Hyperlink controls associated with
```

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs:81]: if no conditions are met, should we record that PackageA was clicked, but nothing happened?

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Telemetry/HyperlinkClickedTelemetryEventTests.cs
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Telemetry/HyperlinkClickedTelemetryEventTests.cs]: Should this be NuGet.VisualStudio.Internal.Contracts.ItemFilter - and if so, I'd use the `using ContractItemFilter = NuGet.VisualStudio.Internal.Contracts.ItemFilter;`

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NuGet.PackageManagement.VisualStudio.Test.csproj
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NuGet.PackageManagement.VisualStudio.Test.csproj]:
```suggestion
    <Compile Include="Telemetry\SearchSelectionTelemetryEventTests.cs" />
```

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Telemetry/HyperlinkClickedTelemetryEventTests.cs
Comment (dominoFire) [reply test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/Telemetry/HyperlinkClickedTelemetryEventTests.cs]: Changed all instnaces of ContractItemFilter

File: src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs
Comment (dominoFire) [reply src/NuGet.Clients/NuGet.PackageManagement.UI/Xamls/DetailControl.xaml.cs:81]: No right now, because not all Hyperlink controls are tracked. You need to add ad CommandParameter. This `if` conditiional guards agains hyperlink controls without that CommandParameter

**Review by donnie-msft — APPROVED** [review]

**Pull request merged**
